pkg/codegen/python/gen_program.go

```diff
@@ -521,14 +521,19 @@ func (g *generator) genPreamble(w io.Writer, program *pcl.Program, preambleHelpe
 			if pkg == "pulumi" {
 				continue
 			}
-			packageName := "pulumi_" + makeValidIdentifier(pkg)
+			var packageName string
 			if r.Schema != nil && r.Schema.PackageReference != nil {
 				pkg, err := r.Schema.PackageReference.Definition()
 				if err == nil {
 					if pkgInfo, ok := pkg.Language["python"].(PackageInfo); ok && pkgInfo.PackageName != "" {
 						packageName = pkgInfo.PackageName
 					}
 				}
+				if packageName == "" {
+					packageName = pyPack(pkg.Namespace, pkg.Name)
+				}
+			} else {
+				packageName = "pulumi_" + makeValidIdentifier(pkg)
```

Comment:
Do we have test coverage for programgen changes?

Reply:
Yes, it's tested in conformace tests. https://github.com/pulumi/pulumi/pull/18418/files#diff-5238f2223268bfda526a6b239228024e191234dec230d3bfe4ba18dc4566db84R2 this line is generated with programgen unless I'm misunderstanding something.